File: 110.py
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        isbal = True
        def dfs(node):
            if node is None:
                return 0
            ld = dfs(node.left)
            rd = dfs(node.right)
            nonlocal isbal
            if not isbal:
                return 0
            if abs(ld - rd) > 1:
                isbal = False
                return 0
            return max(ld, rd) + 1
        dfs(root)
        return isbal

File: sol/110/110.cpp
struct Solution {
    int dfs(TreeNode *node, bool &isbal){
        if (node == nullptr){
            return 0;
        }
        int ld = this->dfs(node->left, isbal), rd = this->dfs(node->right, isbal);
        if (!isbal){
            return 0;
        }
        if (abs(ld - rd) > 1){
            isbal = false;
            return 0;
        }
        return max(ld, rd) + 1;
    }
    inline bool isBalanced(TreeNode* root) {
        bool isbal = true;
        this->dfs(root, isbal);
        return isbal;
    }
};
